d3b_release_maker/release_maker.py

```diff
@@ -0,0 +1,407 @@
+
+        return version, markdown
+
+
+def new_notes(repo, blurb_file):
+    """
+    Build notes for new changes
+    """
+    blurb = None
+    if blurb_file:
+        with open(blurb_file, "r") as bf:
+            blurb = bf.read().strip()
+
+    return GitHubReleaseNotes().build_release_notes(repo=repo, blurb=blurb)
+
+
+def new_changelog(repo, blurb_file):
+    """
+    Creates release notes markdown containing:
+    - The next release version number
+    - A changelog of Pull Requests merged into master since the last release
+    - Emoji and category summaries for Pull Requests in the release
+
+    Then merges that into the existing changelog.
+    """
+
+    # Build notes for new changes
+
+    new_version, new_markdown = new_notes(repo, blurb_file)
+
+    if new_version not in new_markdown.partition("\n")[0]:
+        print(
+            f"New version '{new_version}' not in release title of new markdown."
+        )
+        return None, None, None
+
+    # Load previous changelog file
+
+    session = GitHubSession()
+    try:
+        prev_markdown = session.get(f"{GH_RAW}/{repo}/master/{CHANGEFILE}").text
+    except UnknownObjectException:
+        prev_markdown = ""
+
+    # Remove previous title line if not specific to a particular release
+
+    if "\n" in prev_markdown:
+        prev_title, prev_markdown = prev_markdown.split("\n", 1)
+        if regex.search(r"[Rr]elease .*\d+\.\d+\.\d+", prev_title):
+            prev_markdown = "\n".join([prev_title, prev_markdown])
+
+    # Update changelog with new release notes
+
+    if new_version in prev_markdown.partition("\n")[0]:
+        print(f"\nNew version '{new_version}' already in {CHANGEFILE}.")
+        return None, None, None
+    else:
+        changelog = "\n\n".join([new_markdown, prev_markdown]).rstrip()
+        return new_version, new_markdown, changelog
+
+
+def make_release(repo, project_title, blurb_file, pre_release_script):
+    """
+    Generate a new changelog, run the script, and then make a PR on GitHub
+    """
+    gh_token = os.getenv(config.GH_TOKEN_VAR)
+
+    new_version, new_markdown, changelog = new_changelog(repo, blurb_file)
+
+    if changelog:
+        # Attach project header
+        changelog = f"# {project_title} Change History\n\n{changelog}"
+
+        # Freshly clone repo
+        tmp = os.path.join(tempfile.gettempdir(), "release_maker")
+        shutil.rmtree(tmp, ignore_errors=True)
+        print(f"Cloning https://github.com/{repo}.git to {tmp} ...")
+        os.system(
+            f"git clone --quiet --depth 1 https://{gh_token}@github.com/{repo}.git {tmp}"
+        )
+        os.chdir(tmp)
+
+        print("Writing updated changelog file ...")
+        with open(CHANGEFILE, "w") as cl:
+            cl.write(changelog)
+
+        if pre_release_script:
+            print(f"Executing pre-release script {pre_release_script} ...")
+            os.chmod(pre_release_script, "u+x")
+            os.system(pre_release_script)
```

Comment:
Will `os.system` raise an exception if something goes wrong? If not, maybe use `subprocess`? Whatever we use, as long as make release fails if the pre-release script fails

Reply:
I guess I have the same comment for the other `os.system` calls too. If they fail, will an exception be raised or will this script keep executing?

Reply:
ok e39398f67f25099e1cd990dd39ce2b9ddd4a4501